### PR TITLE
DAO & Mock Registry

### PR DESCRIPTION
**Pull request by eriknelson**
* Implemented some basic dao features for talking with etcd
* Filled out a mock-registry as a separate binary so we have something to pull data from in a real way.

### CODE REVIEW

File: pkg/ansibleapp/dev_registry.go [hunk 1]
```diff
@@ -1,20 +1,63 @@
...
+	log    *logging.Logger
 }
 
-func (r *DevRegistry) Init(config RegistryConfig) error {
+func (r *DevRegistry) Init(config RegistryConfig, log *logging.Logger) error {
+	log.Debug("RHCCRegistry::Init")
 	r.config = config
-	fmt.Printf("DevRegistry::Init with url -> [ %s ] \n", r.config.Url)
+	r.log = log
 	return nil
```
Comment (jmrodri) [inline pkg/ansibleapp/dev_registry.go]: Why return anything if we're just going to return nil?

File: pkg/ansibleapp/registry.go [hunk 2]
```diff
@@ -22,7 +31,10 @@ func CreateRegistry(config RegistryConfig) Registry {
...
 	}
 
-	reg.Init(config)
+	err := reg.Init(config, log)
+	if err != nil {
+		return nil, err
+	}
 
-	return reg
+	return reg, err
```
Comment (jmrodri) [inline pkg/ansibleapp/registry.go]: Looks good.

File: pkg/ansibleapp/rhcc_registry.go [hunk 3]
```diff
@@ -1,20 +1,22 @@
...
 	return nil
 }
 
-func (r *RHCCRegistry) LoadApps() error {
-	fmt.Println("RHCCRegistry::LoadApps ")
-	return nil
+func (r *RHCCRegistry) LoadApps() ([]*Spec, error) {
+	r.log.Debug("RHCCRegistry::LoadApps")
+	return []*Spec{}, nil
 }
```
Comment (jmrodri) [inline pkg/ansibleapp/rhcc_registry.go]: I do like the pluggable registry mechanism.

File: pkg/app/app.go [hunk 4]
```diff
@@ -1,24 +1,37 @@
...
+	registry ansibleapp.Registry
 }
 
 func CreateApp() App {
 	var err error
+
+	fmt.Println("============================================================")
+	fmt.Println("==           Starting Ansible Service Broker...           ==")
+	fmt.Println("============================================================")
+
```
Comment (jmrodri) [inline pkg/app/app.go]: NICE formatting. :tada: :trophy: :medal_sports:

File: pkg/app/app.go [hunk 5]
```diff
@@ -36,11 +49,28 @@ func CreateApp() App {
...
+		os.Exit(1)
+	}
+
 	return app
 }
 
 func (a *App) Start() {
-	a.log.Info("Starting application")
+	a.log.Notice("Ansible Service Broker Started")
+	a.registry.LoadApps()
```
Comment (jmrodri) [inline pkg/app/app.go]: What if this fails?

File: pkg/dao/dao.go [hunk 6]
```diff
@@ -1,10 +1,108 @@
...
+	}
+
+	return nil
+}
+
+////////////////////////////////////////////////////////////
+// Key generators
+////////////////////////////////////////////////////////////
+func specKey(id string) string {
+	return fmt.Sprintf("/spec/%s", id)
```
Comment (jmrodri) [inline pkg/dao/dao.go]: This file looks good too.

File: pkg/ansibleapp/dev_registry.go (hunk 1 above)
Comment (eriknelson) [inline pkg/ansibleapp/dev_registry.go]: Won't compile without it

Comment (eriknelson) [inline pkg/ansibleapp/dev_registry.go]: I expect real registries to be doing more interesting things with failure possibility; this guys just kind of a test stub for us for now.

File: pkg/app/app.go (hunk 5 above)
Comment (eriknelson) [inline pkg/app/app.go]: Good catch; I actually should remove this line for this PR. I'm picturing this start method being the one responsibly for starting the http mux listener. That will call into methods like `broker.Bootstrap()` (not written yet), which *in turn*, will be the guy that calls `registry.LoadApps()`.

So the full story has yet to be written here. This PR should be a lot of support without action in `Start()`. I'll try to make a note of this in `//TODOs` to explain myself a little better inline.

File: pkg/ansibleapp/dev_registry.go (hunk 1 above)
Comment (jmrodri) [inline pkg/ansibleapp/dev_registry.go]: Oh I get it now. The interface signature requires an error for REAL registries. For the dev one we don't care. Ok I forgot about that. +1

**Comment by jmrodri** [comment]
ACK

**Pull request merged**